presto-delta/src/main/java/com/facebook/presto/delta/DeltaPageSourceProvider.java

```diff
@@ -0,0 +1,409 @@
+import static com.facebook.presto.parquet.ParquetTypeUtils.getSubfieldType;
+import static com.facebook.presto.parquet.ParquetTypeUtils.lookupColumnByName;
+import static com.facebook.presto.parquet.ParquetTypeUtils.nestedColumnPath;
+import static com.facebook.presto.parquet.predicate.PredicateUtils.buildPredicate;
+import static com.facebook.presto.parquet.predicate.PredicateUtils.predicateMatches;
+import static com.facebook.presto.spi.StandardErrorCode.PERMISSION_DENIED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.nullToEmpty;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.parquet.io.ColumnIOConverter.constructField;
+import static org.apache.parquet.io.ColumnIOConverter.findNestedColumnIO;
+
+public class DeltaPageSourceProvider
+        implements ConnectorPageSourceProvider
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final FileFormatDataSourceStats fileFormatDataSourceStats;
+
+    @Inject
+    public DeltaPageSourceProvider(
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager,
+            FileFormatDataSourceStats fileFormatDataSourceStats)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.fileFormatDataSourceStats = requireNonNull(fileFormatDataSourceStats, "fileFormatDataSourceStats is null");
+    }
+
+    @Override
+    public ConnectorPageSource createPageSource(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableLayoutHandle layout,
+            List<ColumnHandle> columns,
+            SplitContext splitContext)
+    {
+        DeltaSplit deltaSplit = (DeltaSplit) split;
+        DeltaTableLayoutHandle deltaTableLayoutHandle = (DeltaTableLayoutHandle) layout;
+        DeltaTableHandle deltaTableHandle = deltaTableLayoutHandle.getTable();
+
+        HdfsContext hdfsContext = new HdfsContext(
+                session,
+                deltaSplit.getSchema(),
+                deltaSplit.getTable(),
+                deltaSplit.getFilePath(),
+                false);
+        Path filePath = new Path(deltaSplit.getFilePath());
+        List<DeltaColumnHandle> deltaColumnHandles = columns.stream()
+                .map(DeltaColumnHandle.class::cast)
+                .collect(Collectors.toList());
+
+        List<DeltaColumnHandle> regularColumnHandles = deltaColumnHandles.stream()
+                .filter(columnHandle -> columnHandle.getColumnType() != PARTITION)
+                .collect(Collectors.toList());
+
+        ConnectorPageSource dataPageSource = createParquetPageSource(
```

Comment:
Technically it is possible to have ORC, but as far as I know currently only the Parquet is supported now. We have check in DeltaClient.java when loading a Delta table to make sure the format is Parquet, otherwise we throw an unsupported error. Later when ORC support is added, we can change this code.